test/rules-custom.spec.ts

```diff
@@ -1,72 +1,87 @@
 import 'mocha';
 
 import { DataLayerRule } from '../src/observer';
-import {
-  expectEqual, waitForFS, setupGlobals, ExpectObserver, expectGlobal, expectNoCalls,
-} from './utils/mocha';
+import { expectEqual } from './utils/mocha';
+import { RulesetTestHarness, getRulesetTestEnvironments } from './utils/ruleset-test-harness';
+import { AppMeasurement } from './mocks/adobe';
 
-describe('Custom rules', () => {
-  afterEach(() => {
-    ExpectObserver.getInstance().cleanup();
-  });
+declare global {
+  // eslint-disable-next-line no-var, vars-on-top
+  var s: AppMeasurement;
+}
 
-  it('should only trigger rules for property changes that match rule selector', async function test() {
-    // Extend test timeout to accommodate waits for event dispatching
-    (this as any).timeout(3000);
+describe('Ruleset: Custom rules', () => {
+  getRulesetTestEnvironments().forEach((testEnv) => {
+    describe(`test environment: ${testEnv.name}`, () => {
+      let testHarness: RulesetTestHarness;
 
-    const s = {
-      eVar11: 'test',
-      eVar22: 'test',
-    };
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
 
-    setupGlobals(([
-      ['s', s],
-    ]));
+      after(async () => {
+        await testEnv.tearDown();
+      });
 
-    const rules: DataLayerRule[] = [
-      {
-        source: 's[^(eVar1)]',
-        operators: [
+      it('only triggers rules for property changes that match rule selector', async () => {
+        const rules: DataLayerRule[] = [
           {
-            name: 'insert',
-            value: 'eVar1',
+            source: 's[^(eVar1)]',
+            operators: [
+              {
+                name: 'insert',
+                value: 'eVar1',
+              },
+            ],
+            destination: 'FS.event',
           },
-        ],
-        destination: 'FS.event',
-      },
-      {
-        source: 's[^(eVar2)]',
-        operators: [
           {
-            name: 'insert',
-            value: 'eVar2',
+            source: 's[^(eVar2)]',
+            operators: [
+              {
+                name: 'insert',
+                value: 'eVar2',
+              },
+            ],
+            destination: 'FS.event',
           },
-        ],
-        destination: 'FS.event',
-      },
-    ];
+        ];
 
-    ExpectObserver.getInstance().create({ rules });
+        testHarness = await testEnv.createTestHarness(rules, {
+          s: {
+            eVar11: 'test',
+            eVar22: 'test',
+          },
+        });
 
-    expectGlobal('s').eVar11 = 11;
+        await testHarness.execute(() => {
```

Comment:
this test technically doesn't follow AAA pattern, i do think that's okay here but the way these tests are organized makes it look like two separate tests in the AAA pattern. no action to be taken, just a brain dump.

Reply:
True; we have more of an Arrange -> Act -> Assert -> Act -> Assert flow here. I find myself leaning into flows like this more in integration test scenarios where multiple actions build upon each other to demonstrate the target behavior. In this case, we previously would've triggered both rules when editing a property matching either rule source.

We could update this test to perform both actions at once (setting `eVar11` and `eVar12`) -- then reversing our assertions since the `FS.event` call queue is currently treated like a stack. LMK if you think that would be clearer or if you have other ideas for how to make this test less surprising while also clearly verifying the intended behavior -- setting a property value matching one rule source doesn't trigger another rule on the same base object (`s` in this case) whose source doesn't match the changed property.